### ISSUE TITLE
Work around caches

### ISSUE THREAD

**Issue by mylaluna** [keke8273/Pikeman#271]
# Background
At the moment, Pikeman caches calculation results so that users do not need to wait to get the same result in display. However, there are a few situations that we need to clear the cache and recalculate everything.
For example:
-  Update an exam;
-  Update teacher assignment info;
-  Update class info;
-  User clicks on recalculate button (Refresh button)

# Requirement
Make sure caches get cleared in above situations.

**Comment by keke8273** [comment]
Clear result cache for a specific exam when teacher assignment and class info is updated is non-trivial. In these situation, the user will need to press the refresh button. 

Update an exam will automatically clear the result in the cache if the result exist.

**Comment by mylaluna** [comment]
Does it sounds too doggy to clear all cache for exams instead of a specific one.

**Comment by keke8273** [comment]
When you update an exam, only the updated exam's result cache will be cleared. 
When you press the refresh button, only the selected exam's result cache will be cleared. 

What is the doggy here?

**Comment by mylaluna** [comment]
Just try to minimise the usage of refresh button. Because users need to think when to use refresh button. Think about when you will use the reload button on website. 
When something goes wrong obviously, right? Therefore, we need to teach users to identify when things go wrong and then press refresh button. It makes our design/process looks bad as it can guide users to a wrong path. If clear all cache is non-trivial, do not worry about it.

**Comment by keke8273** [comment]
We are talking two different issues there. That is why I asked for two issues in the first place not one. 

1. I agree to minimize the use of refresh. Blow away the entire cache when a teacher assignment is updated is easy. 

2. I still insist on clearing a specific cache when an exam is updated. There is no need to blow away other exams' result if they are not affected.

**Comment by mylaluna** [comment]
From design point of view, I feel these two are the same to me.
1. Glad to know you agree about the usage of refresh and clear cache is easy.
2. I agree it would be nicer to just clear a specific cache but you said it is non-trival. Given the time and freebie situation, would clear all too doggy being a compromising/temporary method.

**Comment by keke8273** [comment]
I didnt make my clear before. All the cached result is tagged with the exam Id. so we update an exam, the exam is not changed, only the test result values are updated. So it is easy to identify a cached result in the cache and remove it. 

What I meant to say was, when you update a teacher assignment, it is difficult to establish a link between a teacher assignment change to an exam or may a few exams. It is possible but difficult to implement. Therefore, I was proposing manual refresh by the user after a teacher assignment is updated. Now we agree on minimizing the use of refresh button, the clear solution is to blow away the entire cache when a teacher assignment is update, because it is hard to pin point the exam(s) that is affected by this change. 

If we are blowing away the entire cache, I started to thing, we dont need a refresh button anyway. Because, the exam is updated, its cache will cleared automatically. If the teacher assignment is updated, the entire cache is cleared. I cant think of any other scenario that would require a refresh.

**Comment by mylaluna** [comment]
I agree above points. We do not need refresh button if we can cover all scenarios that could calculation accuracy.
